packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ClientOptionsProvider.cs

```diff
@@ -6,25 +6,128 @@
 
         public ClientOptionsProvider(InputClient inputClient)
         {
             _inputClient = inputClient;
+            _clientProvider = ClientModelPlugin.Instance.TypeFactory.CreateClient(inputClient);
+
+            var inputEnumType = ClientModelPlugin.Instance.InputLibrary.InputNamespace.Enums
+                    .FirstOrDefault(e => e.Usage.HasFlag(InputModelTypeUsage.ApiVersionEnum));
+            if (inputEnumType != null)
+            {
+                _serviceVersionEnum = ClientModelPlugin.Instance.TypeFactory.CreateEnum(inputEnumType, this);
+                _latestVersionField = new(
+                    modifiers: FieldModifiers.Private | FieldModifiers.Const,
+                    type: _serviceVersionEnum.Type,
+                    name: LatestVersionFieldName,
+                    initializationValue: Static(_serviceVersionEnum.Type).Property(_serviceVersionEnum.EnumValues[^1].Name));
+                _versionProperty = new(
+                    null,
+                    MethodSignatureModifiers.Internal,
+                    typeof(string),
+                    VersionPropertyName,
+                    new AutoPropertyBody(false));
+            }
         }
 
         protected override string BuildRelativeFilePath() => Path.Combine("src", "Generated", $"{Name}.cs");
-        protected override string BuildName() => $"{_inputClient.Name}Options".ToCleanName();
+        protected override string BuildName() => $"{_clientProvider.Name}Options";
+        protected override FormattableString Description => $"Client options for {_clientProvider.Type:C}.";
 
         protected override CSharpType[] BuildImplements()
         {
             return [typeof(ClientPipelineOptions)];
         }
+
+        protected override FieldProvider[] BuildFields()
+        {
+            if (_latestVersionField == null)
+                return [];
+
+            return [_latestVersionField];
+        }
+
+        protected override TypeProvider[] BuildNestedTypes()
+        {
+            if (_serviceVersionEnum == null)
+                return [];
+
+            return [_serviceVersionEnum];
+        }
+
+        protected override ConstructorProvider[] BuildConstructors()
+        {
+            if (_serviceVersionEnum == null || _latestVersionField == null)
+                return [];
+
+            var versionParam = new ParameterProvider(
+                "version",
+                $"The service version",
+                _serviceVersionEnum.Type,
+                defaultValue: _latestVersionField);
+            var serviceVersionsCount = _serviceVersionEnum.EnumValues.Count;
+            List<SwitchCaseExpression> switchCases = new(serviceVersionsCount + 1);
+
+            for (int i = 0; i < serviceVersionsCount; i++)
+            {
+                var serviceVersionMember = _serviceVersionEnum.EnumValues[i];
+                // ServiceVersion.Version => "version"
+                switchCases.Add(new(
```

Comment:
nit: when we have to new expressions its usually a code smell on our expression language.

Can you use this instead
```c#
switchCases.Add(new(
    Static(_serviceVersionEnum.Type).Property(serviceVersionMember.Name),
    Literal(serviceVersionMember.Value)));
```